### ISSUE TITLE
Refactoring of sorting logic

### ISSUE THREAD

**Issue by tesshucom** [tesshucom/jpsonic#1938]
Related #1925. ~~Prerequisites: #1879~~.  Fixed prior to #1879. Comparator modifications, also mentioned in this issue, will be made when the index design changes.

A refactoring on sorting logic will be done. 

### Overview

Many tables now have a default Order column. By taking advantage of this, some dynamic sorting that occurs during browsing could be bypassed. It's a bit unlikely that this would dramatically improve response. However, it is desirable to keep things tidy in order to provide new sorting options.

<details>
<summary>Background</summary>
The Order column was introduced quite some time ago. However, due to performance considerations, it was optional prior to v112. Since v112 it is mandatory and will be refleshed during scanning.

 - As a first workaround before v112, Jpsonic unified various sorting logics into a common sorting logic. The old Subsonic family included sorting that was dependent on the database implementation and had inconsistent sorting implementations. Specifically, depending on the database employed, the order of uppercase A and lowercase a may be reversed. (We may not care unexpectedly, but the sort order implementation is  differently for each database. You can change the settings each db.) Or, the order differs depending on the view to be displayed.  It means that consistency is not taken even if it is limited to the alphabet. 
 - As a second workaround, in v112, creating a sort key when scanning is mandatory. There's a reason to do it even if it makes the scan cost. In Jpsonic, accuracy was obtained in exchange for a certain cost. Or partial scans are possible in the future (It's never just a disadvantage).
   - A fetch is required to do an accurate delta update. Accurate fetch requires a unique sort order. Legacy servers cannot do this, so processing is inconsistent, especially when data duplication exists in some fields. 
   - Legacy server scans rely on sorting of filesystems, so scanning results may differ between platforms.
   - There will be cases where dynamic sorting is virtually impossible. This idea must be accepted. 'Album all' or 'Artist all' exists in the API. Even if your library has hundreds of thousands of songs. In particular, UPnP requires accurate counts and offsets, and DLNA will also require response speed. Some processing will make sense to rely on order columns in clean SQL.

</details>

Sorting accuracy will have a lot to do with multilingual support. And app with inconsistent sorting looks less acceptable to screen reader users.

<details>
<summary>Why Re-Review Is Needed Now</summary>

 - Before and after v112, whether the Order column is required is different. Deeply related parts have been rewritten, but there may be parts that are not.
 - Maintenance may be inadequate due to the large number of. For example, View had Web Pages, REST, UPnP, but sometimes had duplicate methods. (Not a bug, but badly designed. Among such methods, it is better to unify those that are particularly related to order.) 
 - Features that are not used often have low maintenance frequency. There might be something missing.

</details>
<details>
<summary>Goal</summary>

- 1. Check if the conventional SQL-dependent sorting is correct.
  - Will be modified respectively
  - New issues will be issued if anything is found to be too troublesome
- 2. Where existing dynamic sorting is performed by Java, modify it to use it if it can be processed by SQL.
  - Or clarify where it can't be done.

</details>
<details>
<summary>Non-Goal</summary>

 - Not intended to improve speed
 - It is not intended to add new user-selectable sorting options that are not currently provided. However, it is possible that they will do so in the future. If such a new additional design will be made, it will be necessary to complete this task as a prerequisite.

</details>

___

<details>
<summary>Investigation</summary>

 - [x] List the methods where order by is used in Dao.
   - Pick out areas that need fixing 
   - Will be modified respectively. New issues will be issued if anything is found to be too troublesome
 - [x] Review where comparators are used in services and controllers.

</details>
<details>
<summary>the methods where order by is used in Dao</summary>

No | Dao Class | Method | Fields used | Need to fix
-- | -- | -- | -- | --
1 | AlbumDao | getAlbumsForArtist(String, List<MusicFolder>) | album_order |  
2 |   | getAlbumsForArtist(long, long, String, boolean, List<MusicFolder> musicFolders) | (conditional) year, album_order | Y
3 |   | getAlphabeticalAlbums | (conditional) artist_reading, album_order, name_reading, album_order, name_reading | Y
4 |   | getMostFrequentlyPlayedAlbums | play_count |  
5 |   | getMostRecentlyPlayedAlbums | last_played |  
6 |   | getNewestAlbums | created |  
7 |   | getStarredAlbums | created |  
8 |   | getAlbumsByYear | year | Y 
1 | ArtistDao | getAlphabetialArtists | artist_order |  
2 |   | getStarredArtists | created |  
1 | MediaFileDao | getMediaFile(MediaType, long, long , List<MusicFolder>) | media_file_order | Y
2 |   | getChildrenOf | (conditional) year, media_file_order | Y
3 |   | getChildrenWithOrderOf | media_file_order |  
4 |   | getFilesInPlaylist(int) | id |  
5 |   | getFilesInPlaylist(int, long, long) | id |  
6 |   | getSongsForAlbum(String, String) | disc_number, track_number |  
7 |   | getSongsForAlbum(long, long, MediaFile) | track_number |  
8 |   | getSongsForAlbum(long, long, String, String) | track_number |  
9 |   | getVideos | title |  
10 |   | getGenreCounts | name |  
11 |   | getMostFrequentlyPlayedAlbums | play_count |  
12 |   | getMostRecentlyPlayedAlbums | last_played |  
13 |   | getNewestAlbums | created |  
14 |   | getAlphabeticalAlbums | (conditional) media_file_order |  
15 |   | getAlbumsByYear | year | Y
16 |   | getAlbumsByGenre | media_file_order |  
17 |   | getFetchedFirstChildOf | media_file_order |  
18 |   | getSongsByGenre | media_file_order, track_number |  
19 |   | getStarredAlbums | created |  
20 |   | getStarredDirectories | created |  
21 |   | getStarredFiles | created |  
22 |   | getArtistAll | media_file_order |  
23 |   | getRandomSongsForAlbumArtist | media_file_order, album_artist, album | Y
24 |   | getSortOfAlbumToBeFixed | id |  
25 |   | guessPersonsSorts | name |  
26 |   | RandomSongsQueryBuilder#build | rand() | Y
1 | MusicFolderDao | getAllMusicFolders | enabled, folder_order |  
1 | PodcastDao | getEpisodes | publish_date |  
2 |   | getNewestEpisodes | publish_date |  
1 | RatingDao | getHighestRatedAlbums | avg(rating) |  
1 | StaticsDao | getScanLog | start_date |  
2 |   | deleteBefore | start_date |  
3 |   | deleteOtherThanLatest | start_date |  
4 |   | isfolderChangedSinceLastScan | start_date |  
5 |   | getScanEvents | executed |  
6 |   | getLastScanEventType | executed |  
7 |   | getLastScanAllStatuses | start_date |  

</details>
<details>
<summary>Comparators</summary>

Here are the places where Comparator is currently used, i.e. where some sort of dynamic sorting is done:

![image](https://github.com/tesshucom/jpsonic/assets/27724847/dbd92a5c-1152-4d4d-ad75-f0639c867202)

I checked and these are required by design for now.

 - Jpsonic has actively maintained 1,2,3,4, but 5,6,7 have not been updated much.
   - Playlist may be expanded, in the future
   - ajax sorting is sloppy, but not high priority considering frequency of use. (We may replace this with a different function instead of maintaining it) **Premature optimization is the root of all evil**. Maintain status quo for a while.
   - The logic used in SortableArtist in 7 is the one used in the index. It will be removed in [v113.0.0](https://github.com/tesshucom/jpsonic/milestone/69).

About half of these are expected to continue to change in the future. If anything, the new support for sorting podcasts and sorting items in the UPnP root node would be preferred. So it's all transitional.

What should be fixed now is the internal logic of 2 and 3.

 - Comparators in the "1" group are just used during registration, so you'll need access to some fields involved in meta processing.
 - Comparators in the '2' and '3' groups  are only used when searching. Especially for alphabetical sorting, it may be sufficient to refer to the Order field instead of string sorting.

</details>

___

#### Tasks

At the moment the following small refactorings will be done. These fixes do not affect operation. (However, if we leave these things too long, the intent of the source code becomes difficult to read.) It's just cleaning.

After all, it looks like it's just a small fixes of Dao 😄

 - [x] Remove unnecessary overloads and standardize methods
   - Fix AlbumDao#getAlbumsForArtist
   - MediaFileDao#getFilesInPlaylist
   - MediaFileDao#getSongsForAlbum
   - MediaFileDao#getAlbumCount
   - MediaFileDao#resetLastScanned
 - [x] Remove unnecessary order by
   - Fix MediaFileDao#getMediaFile(MediaType, long, long , List)
     - Used only for Windows Media Player dump. Since it is an iterative process, not sorting can speed things up. If you dump during scanning... Hmm. If anything, it might be better not sorting.
   - Fix getRandomSongsForAlbumArtist
     - Since it will be replaced with random anyway, there is no need for sorting
 - [x] Fix year sorting better
   - MediaFileDao#getChildrenOf (year, media_file_order)
     - Year or Alphabetical switching is done by the flag. **year or media_file_order**
     - A better example would be: **year is null, year (desc), media_file_order or media_file_order**
   - Applies to all methods that sort by year
     - AlbumDao#getAlbumsForArtist
     - AlbumDao#getAlbumsByYear
     - MediaFileDao#getChildrenOf
     - MediaFileDao#getAlbumsByYear
 - [x] Standardize methods and remove unnecessary methods (not triaged)
   - ~~Fix MediaFileDao#getChildrenWithOrderOf~~  I just remembered why getChildrenOf bothers to have some overriding methods🥱 It's not unorganized complexity, it's a variety of pitches.
     - from media_file **(No condition)**
     - from media_file **(where ~)**
     - from media_file **(where ~ limit ~ offset ~)**
   - They differ in execution cost. There are many places where it is used, and the number of iterations is also large. Since it has a large impact on performance, it was intentionally divided so that the reference destination can be easily dereferenced. (If the method paramaters were standardized, it would be quite difficult to see)
   - It seems that this place can be left as it is. **Premature optimization is the root of all evil** There is also a faster method than limit/offset if the database type can be limited. If there is an option to partially introduce Dialect in future improvements, the current implementation is better than standardization. getChildren is a bit special.
   - For methods other than getChildren, I don't think there is a big problem if you share the Paramater and make the method common without worrying too much about it. As far as I can see. There is no override-method using SELECT where cost calculation is so fatal. 
 - [x] Remove the inappropriate function (rand()) and process it on the Java side (not triaged)
   - Fix MediaFileDao$RandomSongsQueryBuilder#build
   - ~~But it's not fixed now. Issue new issue.~~ 
     - Oops!  Already existed. #364 
     - The shuffle used by UPnP is already available on the cross platform. Therefore, if rewrite it in a similar way, it will work on any DB. The point is not there, but that advanced search should be redesigned as a whole (hence the neglect until now). If you prepare an alias for rand() in Postgres, it will work, so it will still be left alone. It will be improved collectively when the advanced search is fixed.
 - [x] Fix the field referenced when comparing alphabetical of JpMediaFileComparator (not triaged)
   - ~~For normal GET, use media_file_order instead of **reading-fields~~ (close to impossible)
   - (Currently) the Order field is registered with the following rules 
     - Alphabetical. Articles, Normalization, Sequential Alpha-Number and linguistic features (Japanese) are considered.
       -  artist#artist_order
       -  album#album_order
       -  media_file#media_file_order&&type=DIRECTORY
       -  media_file#media_file_order&&type=ALBUM
     - **TrackNumber, (Path collator comparison if not available)**
       -  media_file#media_file_order&&AudioFile
   - An important use is primarily fetching. Or sort all Artists and Albums. **The reason Alphabestical sort key generation is not done in media_file MUSIC, etc. is Cost for string parsing.** If we have a strategy of forcing the use of the Order field on all SELECT, we'll end up with longer scan times. And there really isn't much motivation for sorting music alphabetically by song. is intentionally omitted
   - Intuitively, replacing getChildrenOf and getMultiFolderChildren with SQL won't make much of a speed difference. Because ...
     - When Children is requested (view ofdirectory children), the number of children is not that many. TimSort will be used in the JDK if the number of children is somewhat high. Sorting in Java isn't all that slow compared to SQL, perhaps for this use case.
     - Songs aren't the only children of directories. Jpsonic's Mediafile Comparator considers sorting in mixed Artist/Album/Songs situations. And there will also be sorting criteria that can be specified on the settings page. As a whole, the cases are relatively complicated, so replacing them with SQL would reduce maintainability.
   - In other words, considering various things, the current implementation is probably good as it is

**Comment by tesshucom** [comment]
I wrote these code a long time ago, so I forgot...

All the parts that have not been replaced with SQL at present have their own reasons.

#### MediaFileService#getChildrenOf


> Roughly speaking, Mediafile, Artist, and Album already have sort keys, so they can be replaced with SQL. 

This perception is incorrect. Mediafile is creating sort keys for directories only. The file is left at the default value of -1.

The comparator used in MediaFileService#getChildrenOf is fairly versatile and can consider many conditions and arrange directories and files in a unique order. Of course we can port it to SQL, but in that case, we need to create a sort key for all files.　(Alphabetical sorting in SQL without using a sort key is NG according to Jpsonic's requirements.)

 - No functional requirement to sort all files (not directories) in the library
 - There may be limited file sorting of certain ranges. Playlists, directories, play queues, etc. However, it does not justify the cost of creating and updating sort keys for all files in the library. Waste of scanning time.
 - In actual use, there is almost no opportunity to arrange music files in a directory alphabetically. Because most of the time it sorts by track number.
 - Certainly, there are cases where neither the year nor the track number is used.  Jpsonic has a limited option "do not sort by year in directories with prefix Various". Probably enough to use it there. (But the child will be a directory, not a file, so a sort key is used already.)


Therefore, the current implementation is almost final form.


#### MainController#getMultiFolderChildren

Possibly unused, dead code. As far as I can see. (I'm not 100% sure as controller tests were almost non-existent.) If anything, the direction of erasing is appropriate, but more important issues should be prioritized.

___

So, for the time being, it should be fine as it is. If sorting features increase and options are added, I think it's safe to remove some non-essential options.

It's probably more profitable to start things like indexing faster first.

**Comment by tesshucom** [comment]
Now, after #1925, the story changes greatly. Ideally, you should use only SQL by default as much as possible. And it is now possible to consider moving to such a design safely.

 - After #2041, Order is registered in all MediaFiles when scanning. This means that all data entries have a default order, including ID3. Data can be retrieved in the same default order using only SQL, regardless of database specifications.
 - Since #1941 MediaFileService and WritableMediaFileService are completely separated.
   - Of course, dealing with sorting on registration is obviously complicated. However, after registration, all you need to know is the order in which they are registered in the Order.
   - MediaFileService's SELECT is fairly safe to rewrite. 

Reconsider!
